File: node/cogs/party_events.js
const { EmbedBuilder } = require('discord.js');

const MAX_REACTIONS = 4;

function buildPartyEmbed({ reasonPt, reasonEn, startTs, endTs, creatorName, members }) {
  const slots = [...members, ...Array(Math.max(0, 5 - members.length)).fill('—')];

  return new EmbedBuilder()
    .setTitle('🎉 Party')
    .setDescription(`**${reasonPt}**\n*${reasonEn}*`)
    .setColor(0x2ecc71)
    .addFields({
      name: '🕒 Disponibilidade / Availability',
      value: `<t:${startTs}:t> – <t:${endTs}:t>`,
      inline: false,
    })
    .addFields({
      name: `👥 Party (${members.length}/5)`,
      value: slots.join('\n'),
      inline: false,
    })
    .setFooter({ text: `Criador: ${creatorName}` });
}

async function updatePartyEmbed({ message, party, members }) {
  const creator = await message.guild.members.fetch(party.creator_id).catch(() => null);
  const creatorName = creator?.displayName || creator?.user?.username || 'Criador';

  const embed = buildPartyEmbed({
    reasonPt: party.reason_pt,
    reasonEn: party.reason_en,
    startTs: party.start_ts,
    endTs: party.end_ts,
    creatorName,
    members,
  });

  await message.edit({ embeds: [embed] });
}

function registerPartyEvents({ client, db }) {
  client.on('messageReactionAdd', async (reaction, user) => {
    if (user.bot) {
      return;
    }

    if (reaction.partial) {
      await reaction.fetch().catch(() => null);
    }
    if (reaction.message.partial) {
      await reaction.message.fetch().catch(() => null);
    }

    if (reaction.emoji.name !== '✅') {
      return;
    }

    const party = await db.getPartyByMessageId(reaction.message.id);
    if (!party) {
      return;
    }

    const users = await reaction.users.fetch();
    const nonBots = users.filter((u) => !u.bot);

    if (nonBots.size > MAX_REACTIONS) {
      await reaction.users.remove(user.id).catch(() => null);
      await user.send('❌ Party cheia (4/4).').catch(() => null);
      return;
    }

    const members = [
      `<@${party.creator_id}>`,
      ...nonBots.map((u) => u.toString()),
    ];

    await updatePartyEmbed({ message: reaction.message, party, members });
  });

  client.on('messageReactionRemove', async (reaction) => {
    if (reaction.partial) {
      await reaction.fetch().catch(() => null);
    }
    if (reaction.message.partial) {
      await reaction.message.fetch().catch(() => null);
    }

    if (reaction.emoji.name !== '✅') {
      return;
    }

    const party = await db.getPartyByMessageId(reaction.message.id);
    if (!party) {
      return;
    }

    const users = await reaction.users.fetch();
    const nonBots = users.filter((u) => !u.bot);

    const members = [
      `<@${party.creator_id}>`,
      ...nonBots.map((u) => u.toString()),
    ];

    await updatePartyEmbed({ message: reaction.message, party, members });
  });
}

module.exports = {
  registerPartyEvents,
};

File: cogs/party_events.py
import discord
from discord.ext import commands

import db
from utils.party_embed import build_party_embed

MAX_REACTIONS = 4


class PartyEvents(commands.Cog):
    def __init__(self, bot):
        self.bot = bot

    @commands.Cog.listener()
    async def on_raw_reaction_add(self, payload):
        if payload.emoji.name != "✅":
            return

        data = db.get_party_by_message(payload.message_id)
        if not data:
            return

        (_, channel_id, creator_id, reason_pt, reason_en, start_ts, end_ts) = data
        channel = self.bot.get_channel(channel_id)
        msg = await channel.fetch_message(payload.message_id)

        reaction = discord.utils.get(msg.reactions, emoji="✅")
        users = [u async for u in reaction.users() if not u.bot]

        if len(users) > MAX_REACTIONS:
            await msg.remove_reaction("✅", payload.member)
            await payload.member.send("❌ Party cheia (4/4).")
            return

        creator = msg.guild.get_member(creator_id)
        members = [creator.mention] + [u.mention for u in users]
        embed = build_party_embed(reason_pt, reason_en, start_ts, end_ts, creator, members)
        await msg.edit(embed=embed)

    @commands.Cog.listener()
    async def on_raw_reaction_remove(self, payload):
        if payload.emoji.name != "✅":
            return

        data = db.get_party_by_message(payload.message_id)
        if not data:
            return

        (_, channel_id, creator_id, reason_pt, reason_en, start_ts, end_ts) = data
        channel = self.bot.get_channel(channel_id)
        msg = await channel.fetch_message(payload.message_id)

        reaction = discord.utils.get(msg.reactions, emoji="✅")
        users = [u async for u in reaction.users() if not u.bot]

        creator = msg.guild.get_member(creator_id)
        members = [creator.mention] + [u.mention for u in users]
        embed = build_party_embed(reason_pt, reason_en, start_ts, end_ts, creator, members)
        await msg.edit(embed=embed)


async def setup(bot):
    await bot.add_cog(PartyEvents(bot))
